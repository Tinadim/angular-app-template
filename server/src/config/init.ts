import 'zone.js/dist/zone-node';
import { resolve } from 'path';
import * as appRoot from 'app-root-path';
import * as dotenv from 'dotenv';
import { registerProvider } from '@app/client-provider';
import * as provider from '@client/src/client-provider';

dotenv.config({ path: resolve(appRoot.path, '.env') });

console.log('<<<<< ROOT', appRoot.path, '>>>>>');

const resolvedPath = require.resolve('../../../client/src/client-provider');
// const resolvedPath = resolve(__dirname, '../../client/src/client-provider');

console.log('ENV\n\n', JSON.stringify(process.env.USE_SSR));

registerProvider(resolvedPath, provider);
