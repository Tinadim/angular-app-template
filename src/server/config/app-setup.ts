import express, { Application, Request, Response } from 'express';
import bodyParser from 'body-parser';
import Router from 'express-promise-router';
import { ngExpressEngine } from '@nguniversal/express-engine';
import { APP_BASE_HREF } from '@angular/common';
import { resolve } from 'path';
import { mount } from '../api';
import { AppServerModule } from '../server-side-rendering/app-server-module';

const distFolder = resolve(__dirname, '../../../dist/client');
const useSsr = process.env.USE_SSR;
const renderer = useSsr
    ? (req: Request, res: Response) => res.render('index', { req, providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }] })
    : (req: Request, res: Response) => res.sendFile(`${distFolder}/index.html`);

export function buildApp() {
    const app = express();
    setupMiddleware(app);
    bootEndpoints(app);
    serveStaticFiles(app);
    return app;
}

function serveStaticFiles(app: Application) {
    app.get('*.*', express.static(distFolder, {
        maxAge: '1y'
    }));

    app.get('*', renderer);
}

function setupMiddleware(app: Application) {
    if (useSsr) {
        app.engine('html', ngExpressEngine({
            bootstrap: AppServerModule,
        }));
    }

    app.set('view engine', 'html');
    app.set('views', distFolder);
    app.use(bodyParser.urlencoded({ extended: true }));
    app.use(bodyParser.json());
}

function bootEndpoints(app: Application) {
    const router = Router();
    mount(router);
    app.use('/api', router);
}
