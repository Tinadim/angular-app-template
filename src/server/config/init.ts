import 'zone.js/dist/zone-node';
import { resolve } from 'path';
import { config as configEnv } from 'dotenv-safe';

const example = resolve(process.cwd(), '.env.example');
configEnv({ example });

console.log('<<<<< USE_SSR', process.env.USE_SSR, '>>>>>');
