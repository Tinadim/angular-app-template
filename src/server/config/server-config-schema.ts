import { $enum } from 'ts-enum-util';

enum EnvVars {
    NodeEnv = 'NODE_ENV',
    Port = 'PORT',
    UseShutDownHandler = 'USE_SHUTDOWN_HANDLER',
    UseSsr = 'USE_SSR'
}

enum ArgVars {
    Port = 'port',
    UseShutDownHandler = 'useShutdownHandler',
    UseSsr = 'useSsr'
}

export const ENV_VARS = $enum(EnvVars).getValues();

export const ARG_VARS = $enum(ArgVars).getValues();

export const DEFAULTS = {
    port: 8080,
    useSsr: true,
    useShutdownHandler: false
};

export const envSchema: object = {
    properties: {
        [EnvVars.Port]: {
            type: 'number',
            default: DEFAULTS.port,
        },
        [EnvVars.UseShutDownHandler]: {
            type: 'boolean',
            default: DEFAULTS.useShutdownHandler,
        },
        [EnvVars.UseSsr]: {
            type: 'boolean',
            default: DEFAULTS.useSsr,
        }
    },
    type: 'object'
};

export const argvSchema: object = {
    properties: {
        [ArgVars.Port]: {
            type: 'number',
        },
        [ArgVars.UseShutDownHandler]: {
            type: 'boolean',
        },
        [ArgVars.UseSsr]: {
            type: 'boolean',
        }
    },
    type: 'object'
};
