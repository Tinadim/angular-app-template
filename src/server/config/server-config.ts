import Ajv from 'ajv';
import { camelCase, cloneDeep, mapKeys, pick } from 'lodash';
import { resolve } from 'path';
import { parse as parseArgv } from 'yargs';
import { envSchema, argvSchema, ENV_VARS, ARG_VARS } from './server-config-schema';

export class ServerConfig {
    readonly distFolderPath: string;
    readonly port: number;
    readonly isProduction: boolean;
    readonly useShutdownHandler: boolean;
    readonly useSsr: boolean;

    constructor() {
        this.distFolderPath = resolve(process.cwd(), './dist/client');

        const mergedConfigs = this.mergeConfigs();

        this.isProduction = process.env.NODE_ENV === 'production';
        this.port = mergedConfigs.port;
        this.useSsr = mergedConfigs.useSsr;
        this.useShutdownHandler = mergedConfigs.useShutdownHandler;
    }

    private mergeConfigs() {
        const ajv = new Ajv({
            allErrors: true,
            coerceTypes: true,
            format: 'full',
            useDefaults: true,
        });
        const relevantEnvVars = pick(process.env, ENV_VARS);
        const argvCopy = pick(cloneDeep(parseArgv()), ARG_VARS);
        this.validateEnv(ajv, relevantEnvVars);
        this.validateArgv(ajv, argvCopy);

        const mappedEnvVars = mapKeys(relevantEnvVars, (value: any, key: string) => camelCase(key));
        return Object.assign({}, mappedEnvVars, argvCopy) as any;
    }

    private validateEnv(ajv: Ajv.Ajv, envVars: object) {
        const valid = ajv.validate(envSchema, envVars);
        if (!valid) {
            throw this.formatAjvError(ajv.errors);
        }
    }

    private validateArgv(ajv: Ajv.Ajv, argVars: object) {
        const valid = ajv.validate(argvSchema, argVars);
        if (!valid) {
            throw this.formatAjvError(ajv.errors);
        }
    }

    private formatAjvError(errors: Ajv.ErrorObject[]) {
        // TODO add better error formatting
        throw new Ajv.ValidationError(errors);
    }
}

export const serverConfig = new ServerConfig();
