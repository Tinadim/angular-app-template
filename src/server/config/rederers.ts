import { Request, RequestHandler, Response } from 'express';
import { APP_BASE_HREF } from '@angular/common';
import { serverConfig } from './server-config';

const { useSsr, distFolderPath } = serverConfig;

export function renderIndex(req: Request, res: Response) {
    res.render('index', {
        req,
        providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }],
    });
}

export function sendIndex(req: Request, res: Response) {
    res.sendFile(`${distFolderPath}/index.html`);
}

export function getRenderer(): RequestHandler {
    return useSsr ? renderIndex : sendIndex;
}
