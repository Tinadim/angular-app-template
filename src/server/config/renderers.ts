import { Request, RequestHandler, Response } from 'express';
import { APP_BASE_HREF } from '@angular/common';
import { serverConfig } from './server-config';

export function renderIndex(req: Request, res: Response) {
    res.render('index', {
        req,
        providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }],
    });
}

export function sendIndex(req: Request, res: Response) {
    const { distFolderPath } = serverConfig;
    res.sendFile(`${distFolderPath}/index.html`);
}

export function getRenderer(): RequestHandler {
    return serverConfig.useSsr ? renderIndex : sendIndex;
}
