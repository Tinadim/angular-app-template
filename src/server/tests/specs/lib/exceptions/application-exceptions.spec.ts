import {
    ApplicationError, ErrorCode,
    ErrorDetail, InsufficientPrivilegesError,
    InvalidParametersError, NotFoundError, UnauthorizedError, UnexpectedError,
} from '@server/lib/exceptions';
import { serverConfig } from '@server/config/server-config';

jest.mock('@server/config/server-config');

describe('application-exceptions.spec.ts', () => {
    beforeAll(() => {
        // to omit error stack traces from snapshot
        (serverConfig as any).isProduction = true;
    });

    function testError(error: ApplicationError) {
        const errorObject = error.toJson();
        expect(errorObject).toMatchSnapshot();
    }

    describe('#InvalidParametersError', () => {
        it('Should create an InvalidParametersError with the provided error detail', () => {
            const errorDetail = {
                code: ErrorCode.MissingRequiredParameter,
                message: 'Parameter name is required',
                field: 'name',
            } as ErrorDetail;
            const error = new InvalidParametersError(errorDetail);
            testError(error);
        });

        it('Should create an InvalidParametersError with the provided error details', () => {
            const errorDetails = [
                {
                    code: ErrorCode.MissingRequiredParameter,
                    message: 'Parameter firstName is required',
                    field: 'firstName',
                },
                {
                    code: ErrorCode.MissingRequiredParameter,
                    message: 'Parameter lastName is required',
                    field: 'lastName',
                },
            ] as ErrorDetail[];
            const error = new InvalidParametersError(errorDetails);
            testError(error);
        });

        it('Should create an InvalidParametersError with the provided error message', () => {
            const error = new InvalidParametersError([], 'Lorem ipsum si amet');
            testError(error);
        });

        it('Should return the expected http error code', () => {
            const error = new InvalidParametersError([], 'Lorem ipsum si amet');
            const code = error.getHttpStatusCode();
            expect(code).toBe(400);
        });
    });

    describe('#NotFoundError', () => {
        it('Should create a NotFoundError with default properties', () => {
            const error = new NotFoundError();
            testError(error);
        });

        it('Should create a NotFoundError with the provided error message', () => {
            const error = new NotFoundError('Lorem ipsum sit amet');
            testError(error);
        });

        it('Should return the expected http error code', () => {
            const error = new NotFoundError();
            const code = error.getHttpStatusCode();
            expect(code).toBe(404);
        });
    });

    describe('#UnauthorizedError', () => {
        it('Should create an UnauthorizedError with default properties', () => {
            const error = new UnauthorizedError();
            testError(error);
        });

        it('Should create an UnauthorizedError with the provided error message', () => {
            const error = new UnauthorizedError('Lorem ipsum sit amet');
            testError(error);
        });

        it('Should return the expected http error code', () => {
            const error = new UnauthorizedError();
            const code = error.getHttpStatusCode();
            expect(code).toBe(401);
        });
    });

    describe('#InsufficientPrivilegesError', () => {
        it('Should create an InsufficientPrivilegesError with default properties', () => {
            const error = new InsufficientPrivilegesError();
            testError(error);
        });

        it('Should create an InsufficientPrivilegesError with the provided error message', () => {
            const error = new InsufficientPrivilegesError('Lorem ipsum sit amet');
            testError(error);
        });

        it('Should return the expected http error code', () => {
            const error = new InsufficientPrivilegesError();
            const code = error.getHttpStatusCode();
            expect(code).toBe(403);
        });
    });

    describe('#UnexpectedError', () => {
        it('Should create an UnexpectedError from the provided error', () => {
            const error = new Error('Lorem ipsum sit amet');
            const unexpectedError = new UnexpectedError(error);
            testError(unexpectedError);
        });

        it('Should return the expected http error code', () => {
            const error = new Error('Lorem ipsum sit amet');
            const unexpectedError = new UnexpectedError(error);
            const code = unexpectedError.getHttpStatusCode();
            expect(code).toBe(500);
        });
    });
});
