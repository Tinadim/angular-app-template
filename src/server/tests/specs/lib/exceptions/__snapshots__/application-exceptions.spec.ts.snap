// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application-exceptions.spec.ts #InsufficientPrivilegesError Should create an InsufficientPrivilegesError with default properties 1`] = `
Object {
  "code": "accessDenied",
  "errors": undefined,
  "message": "You don't have access to the requested resource",
}
`;

exports[`application-exceptions.spec.ts #InsufficientPrivilegesError Should create an InsufficientPrivilegesError with the provided error message 1`] = `
Object {
  "code": "accessDenied",
  "errors": undefined,
  "message": "Lorem ipsum sit amet",
}
`;

exports[`application-exceptions.spec.ts #InvalidParametersError Should create an InvalidParametersError with the provided error detail 1`] = `
Object {
  "code": "invalidParameters",
  "errors": Array [
    Object {
      "code": "missingRequiredParameter",
      "field": "name",
      "message": "Parameter name is required",
    },
  ],
  "message": "Invalid parameters provided",
}
`;

exports[`application-exceptions.spec.ts #InvalidParametersError Should create an InvalidParametersError with the provided error details 1`] = `
Object {
  "code": "invalidParameters",
  "errors": Array [
    Object {
      "code": "missingRequiredParameter",
      "field": "firstName",
      "message": "Parameter firstName is required",
    },
    Object {
      "code": "missingRequiredParameter",
      "field": "lastName",
      "message": "Parameter lastName is required",
    },
  ],
  "message": "Invalid parameters provided",
}
`;

exports[`application-exceptions.spec.ts #InvalidParametersError Should create an InvalidParametersError with the provided error message 1`] = `
Object {
  "code": "invalidParameters",
  "errors": Array [],
  "message": "Lorem ipsum si amet",
}
`;

exports[`application-exceptions.spec.ts #NotFoundError Should create a NotFoundError with default properties 1`] = `
Object {
  "code": "entityNotFound",
  "errors": undefined,
  "message": "Resource not found",
}
`;

exports[`application-exceptions.spec.ts #NotFoundError Should create a NotFoundError with the provided error message 1`] = `
Object {
  "code": "entityNotFound",
  "errors": undefined,
  "message": "Lorem ipsum sit amet",
}
`;

exports[`application-exceptions.spec.ts #UnauthorizedError Should create an UnauthorizedError with default properties 1`] = `
Object {
  "code": "authenticationRequired",
  "errors": undefined,
  "message": "Authentication required",
}
`;

exports[`application-exceptions.spec.ts #UnauthorizedError Should create an UnauthorizedError with the provided error message 1`] = `
Object {
  "code": "authenticationRequired",
  "errors": undefined,
  "message": "Lorem ipsum sit amet",
}
`;

exports[`application-exceptions.spec.ts #UnexpectedError Should create an UnexpectedError from the provided error 1`] = `
Object {
  "code": "unexpectedError",
  "errors": Array [
    Object {
      "code": "unexpectedError",
      "field": "originalError",
      "message": "Lorem ipsum sit amet",
    },
  ],
  "message": "Unexpected error",
}
`;
